activerecord/lib/active_record/associations/join_dependency.rb

```diff
@@ -114,12 +114,16 @@ def join_constraints(outer_joins, join_type)
           end
         }
 
-        joins.concat outer_joins.flat_map { |oj|
+        joins.concat joins_to_add.flat_map { |oj|
           if join_root.match? oj.join_root
             walk join_root, oj.join_root
           else
             oj.join_root.children.flat_map { |child|
-              make_outer_joins oj.join_root, child
+              if join_type == Arel::Nodes::OuterJoin
+                make_left_outer_joins oj.join_root, child
```

Comment:
I think it is fine to keep it the same as in the line 110. Could you please rebase. I'll merge it to master. I'm not going to backport since even that it is a bug it may cause behavior change in a stable branch for some people.